An improved Leak (Uncollected Objects) Gramplet

prculley opened the pull request:
The recent bug on Glade file memory leak had someone mention the 'Leak' Gramplet.  I started using it and agreed that it was rather hard to learn anything from it.

So I decided to make a couple of small improvements.  I added a column to the main view that lists the line items referrer(s).  This makes it pretty clear that most of the leaks are due to cyclic references (an item refers to another which refers back to the original).  If there are several referrers, they are listed (unless there are too many, then it adds a '...' to the end).  If a referrer is not found in the list, it adds a '*'.

I also added the referrer number (or '****' if none) to the enter/double-click popup referrer list. It still displays everything, even if there were too many for the main screen column.

I noticed that the Leak code itself added referrers to the list, I tried to exclude these from the referrer list.

I did figure out that our DummyDb was staying around in memory, even after it was being closed.  It had a self reference it inherited from gramps.db.base; I changed that by saying 'self.basedb = None' instead of 'self.basedb = self'.  The basedb init code had only two items in it, so I just copied them into dummydb and removed the basedb init call.  A separate commit if we want to use it.

There are a number of other items in the Leak list I cannot figure out; they look like they involve various Gtk objects, so I suspect that there may be some issue with garbage collection and cyclic references...

Everything I read says that python SHOULD be able to eventually garbage collect even this type of thing, however I never saw this fall of the Leak list, even after multiple calls to gc.collect.

sam-m888 commented:
Works on Sidebars and Bottombars 👍 

Then  attempted to test by attaching the leak (**Uncollected Objects**) Gramplet to the **Dashboard** and get the following error:

```
17958: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/grampletpane.py", line 1439, in add_gramplet
    gramplet.pui.update()
  File "/home/grampsdev/aaSandbox/gramps/gramps/gen/plug/_gramplet.py", line 307, in update
    self._generator = self.main()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 116, in main
    self.display()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 203, in display
    del junk, referrer
UnboundLocalError: local variable 'referrer' referenced before assignment

```

prculley commented:
@sam-m888 Should have removed that line; it was an experiment.  Fixed now.
You must have had no uncollected objects in the list when you did your experiment...

sam-m888 commented:
Ok, thanks for the fix.  This should be ready to merge, can you do it
please?

 Yes, probably because I only started Gramps, dismissed the Family Tree
manger dialog, right clicked on the dashboard and selected the gramplet
when the error assistant popped up.

On 5 February 2017 at 01:57, Paul Culley <notifications@github.com> wrote:

> @sam-m888 <https://github.com/sam-m888> Should have removed that line; it
> was an experiment. Fixed now.
> You must have had no uncollected objects in the list when you did your
> experiment...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gramps-project/gramps/pull/345#issuecomment-277451082>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIguKfhTyQqM9koc_dT6RmwLaxAL63mNks5rZJHagaJpZM4L0M4G>
> .
>

sam-m888 commented:
@prculley When you get a moment can you please rebase this? Thank you

prculley commented:
@sam-m888 The bug when used as a sidebar gramplet was not really related to my fix; it was instead an original code bug.  In any event, I rebased and fixed the issue with another commit.

sam-m888 commented:
> @sam-m888 The bug when used as a sidebar gramplet was not really related to my fix; it was instead an original code bug. In any event, I rebased and fixed the issue with another commit.

Thanks for fixing the sidebar issue.

**DB-API** worked so far.

Sorry to say the Dashboard crashed Gramps (and locked the family tree) for me using:

**bsddb**  with an empty family tree

```
(xgrampspy3) [grampsdev@lfs101x-VirtualBox gramps (prculley-leak)]$ ./Gramps.py 
2017-02-13 10:56:56.230: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/grampletpane.py", line 1439, in add_gramplet
    gramplet.pui.update()
  File "/home/grampsdev/aaSandbox/gramps/gramps/gen/plug/_gramplet.py", line 307, in update
    self._generator = self.main()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 116, in main
    self.display()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 189, in display
    if referrer not in junk:
  File "/home/grampsdev/.virtualenvs/xgrampspy3/lib/python3.5/_collections_abc.py", line 622, in __eq__
    return dict(self.items()) == dict(other.items())
  File "/usr/lib/python3/dist-packages/bsddb3/dbshelve.py", line 178, in items
    items = list(self.db.items())
bsddb3.db.DBError: (0, 'DB object has been closed')


(Gramps.py:4385): Gdk-ERROR **: The program 'Gramps.py' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadAlloc (insufficient resources for operation)'.
  (Details: serial 13715 error_code 11 request_code 130 (MIT-SHM) minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap (core dumped)
(xgrampspy3) [grampsdev@lfs101x-VirtualBox gramps (prculley-leak)]$ 


```
**In-Memory** gave the following (and separately was very slow creating the family tree): 

```

(xgrampspy3) [grampsdev@lfs101x-VirtualBox gramps (prculley-leak)]$ ./Gramps.py 

(Gramps.py:4486): Gdk-ERROR **: The program 'Gramps.py' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadAlloc (insufficient resources for operation)'.
  (Details: serial 52176 error_code 11 request_code 130 (MIT-SHM) minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap (core dumped)

```

sam-m888 commented:
@prculley Are you able to recreate/confirm my last errors?

prculley commented:
@sam-m888 Sorry it took a while to look into this; been busy.  After trying to replicate, I decided that it was going to be hard to guess all the possible exceptions that might occur when looking at data that really shouldn't be there anyway (it was on the garbage list because something thought it was no longer needed).

So even though pylint doesn't like 'bare except', I used it anyway.  This seems to stop the issues you were having.

sam-m888 commented:
Thanks for looking into it, slightly better in that I can only reproduce the error below with all three.

I created three empty family tree one of each type(**bsddb** ,  **DB-API** ,  **In-Memory**) ,  and for each type loaded the Family tree and added the Gramplet to the Dashboard, then from the family tree manager  pressed the close button on the family tree manager to unload the family tree and Gramps crashes and locks the family tree (happens more relialby with  **DB-API** ,  **In-Memory** types).

Do you get this error? Or is it specific to Ubuntu 16.04?

Either way I suggest that this be merged as it is an improvement over what exist, and regular users would not be using it.

Thank you
```
(Gramps.py:4686): Gdk-ERROR **: The program 'Gramps.py' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadAlloc (insufficient resources for operation)'.
  (Details: serial 18130 error_code 11 request_code 130 (MIT-SHM) minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap (core dumped)

```

Nick-Hall commented:
> So even though pylint doesn't like 'bare except', I used it anyway.

In general it is not good practice to use a bare except; it can make debugging more difficult.  However, it can be useful in some circumstances.  The pylint warning is still useful.

Don't try to get rid of all pylint messages.  Use your own judgement.  Hopefully most modules will not produce pages of output.

prculley commented:
@sam-m888 I am unable to recreate the Gdk/X-window error.  However, I did find that the test conditions you used created a whole lot of popup windows (50+).  It is possible that this might have hit some x-window limit.

I did not like the popups, so I studied the problem a bit more and found some more fixes/improvements.

The underlying issue is that I am trying to display information about objects that are in the process of being deleted, and under some conditions the attempt is creating 'ReferenceError'.  It seems that the conditions you found exposed a type of object I had not seen before (weakly-referenced).  I tried to bullet-proof the code to deal with these, I think I have succeeded, but you are pretty good at finding 'interesting' test cases.

I'm sure you will let me know if you find something else; so thanks in advance for testing.

sam-m888 commented:
@prculley Can not seem to trigger that last error on ubuntu Thank you.

> I'm sure you will let me know if you find something else; so thanks in advance for testing.

👍 On further testing, created a bsddb family tree and attached the addon to the dashboard. Closed the family tree from the family tree manager. Selected Edit>preferences >Family tree>Database backend: changed it to DB-API and selected close. Then in the Family tree manager create a new family tree and load it. Gramps becomes unresponsive for me and then the Error assistant appears with the error below, but does not crash or lock:


```
104281: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/viewmanager.py", line 1088, in view_changed
    self.__change_page(page_num)
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/viewmanager.py", line 1101, in __change_page
    self.active_page.set_active()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/view/dashboardview.py", line 109, in set_active
    self.widget.set_active()
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/grampletpane.py", line 1492, in set_active
    self.gramplet_map[title].pui.update()
  File "/home/grampsdev/aaSandbox/gramps/gramps/gen/plug/_gramplet.py", line 307, in update
    self._generator = self.main()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 122, in main
    self.display()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 220, in display
    self.model.append((count, ref, str(each)))
TypeError: __str__ returned non-string (type NoneType)


```

prculley commented:
@sam-m888 So I tried again.  I made a few small changes in performance, but it does take time to refresh the display when there are a lot of objects.  To combat this, I set it up to only do a display when the 'Refresh' button is pressed; not just at startup and new tree.  I added a little reminder text at top when this is needed.

In addition, I added a progress meter to give feedback. 

I tried to further bullet-proof against attempts to show weird objects in the midst of being deleted, hope I did better.  Even if I missed something, you should get a print to the console of the issue, rather than a crash dialog.

sam-m888 commented:
@prculley Thanks for your efforts. 👍    Sorry to say my first test:  "With no Family tree open, adding the gramplet to any category(apart from the Dashboard where you can not add any additional gramplets without an active family tree!) sidebar(or bottombar) and then pressing refresh" gives the following error:

```
30836: ERROR: grampsapp.py: line 143: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 249, in apply_clicked
    self.display()
  File "/home/grampsdev/aaSandbox/gramps/gramps/plugins/gramplet/leak.py", line 199, in display
    _('Updating display...'), '', parent=self.gui.gvwin)
AttributeError: 'TabGramplet' object has no attribute 'gvwin'

```

On the Dashboard with an active family tree you can add the gramplet and it works:

![dashboard-uncollected](https://cloud.githubusercontent.com/assets/8924713/26515347/5d8e0e36-42bd-11e7-973d-dfe2d5f5fb3a.png)

But when you close the active family tree on the Dashboard and press the gramplets refresh button twice (for me) Gramps totally crashes and the coredump error is back:

(xgrampspy3) [grampsdev@lfs101x-VirtualBox gramps (prculley-leak)]$ ./Gramps.py 
```

(Gramps.py:4212): Gdk-ERROR **: The program 'Gramps.py' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadAlloc (insufficient resources for operation)'.
  (Details: serial 13091 error_code 11 request_code 130 (MIT-SHM) minor_code 5)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap (core dumped)

```

prculley commented:
@sam-m888 Thanks for your continued testing, we are probably spending too much time on this developer tool...

Regarding the parent problem; I guess I never thought of putting the leak in one of the views... Fixed now.  I have to wonder if any other Gramplets are doing popups, and if they are correctly parented for all these cases (or parented at all), I'm afraid to look...

I have not been able to replicate your x-windows crash.  So no idea what to do next.  Perhaps you could try setting the GDK_SYNCHRONIZE=1 in your environment before running.  Maybe the backtrace will be more useful then.

I also caused the 'Uncollected objects' count at the top of the Gramplet to be drawn first, before starting the  list update.  I'd be curious how many objects you were seeing when it crashed.  I normally get <1000 for my simple open/close tests.  The 'insufficient resources for operation' suggests too much of something...

sam-m888 commented:
> we are probably spending too much time on this developer tool...

@prculley Agreed 👍 

>  have not been able to replicate your x-windows crash. So no idea what to do next. Perhaps you could try setting the GDK_SYNCHRONIZE=1 in your environment before running. Maybe the backtrace will be more useful then.

Experimented attempting to get a backtrace and only successfully(sic) managed to destroy my VM(which I only rebuilt today!) I have a feeling it's more related to me running it in a Virtualbox VM so I'll leave it alone, unless someone can replcate in on a bare PC.

@Nick-Hall Not to sure if I should be merging this PR seeing as you've marked it 5.1?  What should I do?

prculley commented:
@Nick-Hall It is definitely an enhancement, so belongs in 5.1 per our guidelines.  Why would you not want to merge?  Are you leaving it up to Sam since he did testing?  Are we avoiding 5.1 work for now to get out 5.0?  Or some other reason?

Nick-Hall commented:
Now that the gramps50 branch has been created this can be merged into master (which will become v5.1).  It would be helpful if developers prioritised the 5.0 release, but this doesn't mean that 5.1 development should stop.

@sam-m888 Please feel free to merge this when you are ready.

sam-m888 commented:
@prculley @Nick-Hall Was about to merge this but test are broken for some reason, maybe next week. don't have time to check why :(

prculley commented:
@sam-m888 Travis is broken because the most recent commit of the Pro-Gen import did not fix up the imports_test for the changed functionality (he added the current date to his imported data, which cannot match the date the test data was generated).  I've sent @apoettker an email detailing one way to fix it, but he hasn't dealt with it yet.

sam-m888 commented:
@prculley Thanks in case that I will leave alone until test fixed.

> (he added the current date to his imported data, which cannot match the date the test data was generated).

@prculley For the test date would it make more sense to only test for a valid date string in that field, making the test less fragile?

prculley commented:
@sam-m888 I suggested 3 ways for him to deal with this, 1) easiest, change his defaults to not add tags with the current date.
2) use the mock library to patch his plugin so that the date/time functions always return the same date. Also pretty easy, especially since the test already has examples of how that is done.
3) hardest, write a filter to look for where he adds _text_ strings with dates in it and except them.  Which is sort of what you suggest.

I'm pretty busy at the moment, so I thought I would let him deal with his issue, but I'm beginning to think he is not going to...  I'll get to it when I finish up what I'm doing now but it might take a couple of weeks.

sam-m888 commented:
Retested.

Thank you 👍
